keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -193,123 +338,651 @@ each list item is made of:
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: APIUnions
+  - Components depending on the feature gate: kube-apiserver
+
+Request handlers in the api server will call into union validation and
+normalization function from the structured-merge-diff repo when feature is
+enabled.
+
+
+###### Does enabling the feature change any default behavior?
+
+No, since all built in unions are currently validated in other ways.
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
 
-#### Beta -> GA Graduation
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
 
-- CRD support has been proven successful
-- Core-types all implement the semantics properly
-- Stable and bug-free for two releases
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
 
-## Future Work
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
 
-Since the proposal only normalizes the object after the patch has been applied,
-it is hard to fail if the patch is invalid. There are scenarios where the patch
-is invalid but it results in an unpredictable object. For example, if a patch
-sends both a discriminator and a field that is not the discriminated field, it
-will either clear the value sent if the discriminator changes, or it will change
-the value of the sent discriminator.
+Yes, requests will simply skip union validation and normalization.
 
-Validating patches is not a problem that we want to tackle now, but we can
-validate "Apply" objects to make sure that they do not define such broken
-semantics.
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Requests will resume performing union validation and normalization; there is no
```

Comment:
This claim doesn't seem quite right.  When used to enforce validation on CRs, there is persistent state in etcd for CR instances.  Some may have this validation enforced and some may not.

Reply:
Yes, I guess that would cause your CRs to be invalid once you turn union validation back on, but it seems strange to expect to continue to support those? 

You mentioned briefly that ratcheting is not supported in CRs, do you have any ideas about supporting this, otherwise I can only think that we would have to say your CRs become invalid and you have to fix them to continue modifying them, right?